### PR TITLE
coolercontrol: init at 1.1.1

### PR DESCRIPTION
**Pull request by codifryed**
## Description of changes

New package: CoolerControl is a program which controls cooling devices such as AIO coolers and fans under Linux.
https://gitlab.com/coolercontrol/coolercontrol

This package has 2 systemd services, a PySide6 GUI, and is written in both Rust and Python.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by OPNA2608** [comment]
Hmm. The packages build fine & can control my Kraken Z73 (after figuring out the correct order in which the different binaries need to be started), but I'm not sure this
```nix
rec {
  coolercontrold = rustPlatform.buildRustPackage { /*...*/ };
  coolercontrol-liqctld = python3.pkgs.buildPythonApplication { /*...*/ };
  coolercontrol-gui = python3.pkgs.buildPythonApplication { /*...*/ };
}
```
with just a single `callPackage` in `all-packages.nix` is the correct way of doing this. At the very least, the commit title isn't correct enough for the rebuild tags, OfBorg or `nixpkgs-review` to detect any added packages.

**Comment by OPNA2608** [comment]
Closes #245746 btw.

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Could you take a look at https://github.com/OPNA2608/nixpkgs/commit/e38e23b35a44694508c78073c83afb7c3d1f4764 / 
[0001-coolercontrol-Refactor.patch.txt](https://github.com/NixOS/nixpkgs/files/12346968/0001-coolercontrol-Refactor.patch.txt)? I've applied some improvements there to avoid a big wall of code suggestions - commit description describes the main changes. If you'd prefer me to annotate them all here then I can do that as well.

For the systemd system services to work, users will need to add the two daemon derivations to `systemd.packages` in their config. This could also be done in a module so it only requires one option to be flicked on… but I remember seeing complaints awhile ago about modules that are too basic, dunno if there was ever a consensus on that.

Regarding the commit title so automation correctly picks up that there are 3 packages it should test… I'm not sure what's the correct way. I don't see i.e. `eclipses` doing it any differently, so it's prolly fine… Will just manually kick off test builds on CI I guess.

---

I'm currently maintaining our `gkraken` package which has been deprecated in favour of your software. Feel free to add my name to the `maintainers` entry if you're open to co-maintaining these Nix packages. :slightly_smiling_face:

**Comment by codifryed** [comment]
Hey @OPNA2608  thanks for taking a look at this and for the suggestions. Packaging this app isn't the standard affair so I wasn't really sure how best to handle the multiple executables. I took some cues from:

https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/applications/networking/p2p/deluge/default.nix
and
https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/applications/networking/onionshare/default.nix

If you think splitting things up as your patch is the best, then LGTM. Also, thanks for the tip for the systemd.packages.

I was previously a maintainer for gkraken before launching this project, so as far as co-maintaining these packages goes: yes!

File: pkgs/applications/system/coolercontrol/coolercontrol-gui.nix
Comment (codifryed) [inline pkgs/applications/system/coolercontrol/coolercontrol-gui.nix:57]: Out of curiousity, why is this needed/preferred?

Comment (OPNA2608) [reply pkgs/applications/system/coolercontrol/coolercontrol-gui.nix:57]: There were 2 layers of wrappers with that binary: the Python wrapper and the Qt wrapper.

```
↪ env LANG=C ls -ahl /nix/store/54dx6wafi9850ci9kshs3n1s2sqr4mfg-coolercontrol-0.17.0/bin/
total 40K
dr-xr-xr-x 2 root root   96 Jan  1  1970 .
dr-xr-xr-x 6 root root   60 Jan  1  1970 ..
-r-xr-xr-x 1 root root 4.1K Jan  1  1970 ..coolercontrol-wrapped-wrapped
-r-xr-xr-x 1 root root  16K Jan  1  1970 .coolercontrol-wrapped
-r-xr-xr-x 1 root root  16K Jan  1  1970 coolercontrol
```

This just disables the automatic wrapping of the Qt one, and the `preFixup` line consolidates its `makeWrapper` flags into the options for the Python one. Manual entry for this: [When using `wrapGAppsHook` with special derivers you can end up with double wrapped binaries.](https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-common-issues-double-wrapped)

**Comment by OPNA2608** [comment]
@ofborg build coolercontrol.coolercontrold coolercontrol.coolercontrol-liqctld coolercontrol.coolercontrol-gui

File: pkgs/applications/system/coolercontrol/coolercontrol-gui.nix
Comment (codifryed) [reply pkgs/applications/system/coolercontrol/coolercontrol-gui.nix:57]: aha thanks!

**Review by OPNA2608 — CHANGES_REQUESTED** [review]

File: pkgs/applications/system/coolercontrol/coolercontrol-gui.nix
Comment (OPNA2608) [inline pkgs/applications/system/coolercontrol/coolercontrol-gui.nix:65]: Aaaaand we should explicitly set `meta.mainProgram` for these, new expectation by `lib.getExe`:
```
nix-repl> lib.getExe coolercontrol.coolercontrold
trace: warning: getExe: Package "coolercontrold-0.17.0" does not have the meta.mainProgram attribute. We'll assume that the main program has the same name for now, but this behavior is deprecated, because it leads to surprising errors when the assumption does not hold. If the package has a main program, please set `meta.mainProgram` in its definition to make this warning go away. Otherwise, if the package does not have a main program, or if you don't control its definition, specify the full path to the program, such as "${lib.getBin foo}/bin/bar".
"/nix/store/yklmyax42ckzflkpx0w288vp2pz6wpia-coolercontrold-0.17.0/bin/coolercontrold"
```

```suggestion
    mainProgram = "coolercontrol";
  };
```

File: pkgs/applications/system/coolercontrol/coolercontrol-liqctld.nix
Comment (OPNA2608) [inline pkgs/applications/system/coolercontrol/coolercontrol-liqctld.nix:39]:
```suggestion
    mainProgram = "coolercontrol-liqctld";
  };
```

File: pkgs/applications/system/coolercontrol/coolercontrold.nix
Comment (OPNA2608) [inline pkgs/applications/system/coolercontrol/coolercontrold.nix:32]:
```suggestion
    mainProgram = "coolercontrold";
  };
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
I think *now* I'm out of nitpicks, sorry! :sweat_smile:

**Comment by OPNA2608** [comment]
> For the systemd system services to work, users will need to add the two daemon derivations to `systemd.packages` in their config. This could also be done in a module so it only requires one option to be flicked on… but I remember seeing complaints awhile ago about modules that are too basic, dunno if there was ever a consensus on that.

I decided to give this a try and it never launched the services. I think this is https://github.com/NixOS/nixpkgs/issues/81138. In light of this, we should prolly add a module to handle this. I'll look into it.

**Comment by codifryed** [comment]
> I decided to give this a try and it never launched the services on its own. I think this is #81138. In light of this, we should prolly add a module to handle this. I'll look into it.

damn, thanks for taking a look into this. 🙏🏼

**Comment by OPNA2608** [comment]
Can you pull https://github.com/OPNA2608/nixpkgs/commit/161b8dcc9cd090b8b1927f95021d24a221a26d06 into this?

This will offer the following option:
image
which should install `coolercontrol-gui` for the user and set up the `coolercontrol-liqctld` & `coolercontrold` system services, with a workaround for #81138.

**Review by OPNA2608 — APPROVED** [review]
Result of `nixpkgs-review pr 248972` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>coolercontrol.coolercontrol-gui</li>
    <li>coolercontrol.coolercontrol-gui.dist</li>
    <li>coolercontrol.coolercontrol-liqctld</li>
    <li>coolercontrol.coolercontrol-liqctld.dist</li>
    <li>coolercontrol.coolercontrold</li>
  </ul>
</details>

Works on my system, LGTM.
image

**Comment by codifryed** [comment]
Thanks for all the work you put into this @OPNA2608. I know the approval process can take some time, especially for a more complex package like this so getting a second opinion is great. Is there anything I can do to help with this? Should I rebase on current master to keep things up-to-date?

**Comment by OPNA2608** [comment]
I linked this in afew pull request linking channels some time ago. Not sure if there is anything else that can be done to speed this up, aside from periodically re-requesting someone to look at it…

You shouldn't need to rebase unless there's a merge conflict that needs to be solved.

**Comment by OPNA2608** [comment]
Could you update this to [0.17.1](https://gitlab.com/coolercontrol/coolercontrol/-/tags/0.17.1)?

**Review by OPNA2608 — APPROVED** [review]
Result of `nixpkgs-review pr 248972` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>coolercontrol.coolercontrol-gui</li>
    <li>coolercontrol.coolercontrol-gui.dist</li>
    <li>coolercontrol.coolercontrol-liqctld</li>
    <li>coolercontrol.coolercontrol-liqctld.dist</li>
    <li>coolercontrol.coolercontrold</li>
  </ul>
</details>

---

Still works fine for me.

**Comment by tbaumann** [comment]
Fricken amazing! Thanks

How do I use this without pivoting my entire system to your branch of nixpkgs?

I'm still new to nix. This is what I tried.

flake.nix
```nix
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nixpkgs-coolercontrol.url = "github:codifryed/nixpkgs/coolercontrol-0.17.0";
...
```

in my system config:
```nix
{
  inputs,
  config,
  pkgs,
  environment,
  ...
}: {
  imports = [
    inputs.nixos-hardware.nixosModules.common-cpu-amd
    inputs.nixos-hardware.nixosModules.common-gpu-amd
    inputs.nixos-hardware.nixosModules.common-pc-ssd

    ./hardware-configuration.nix
    ../../common/core.nix
    (
      import "${inputs.nixpkgs-coolercontrol}/nixos/modules/programs/coolercontrol.nix" (
        {
          config = config;
          pkgs = inputs.nixpkgs-coolercontrol;
          lib =  inputs.nixpkgs-coolercontrol.lib;
        }
      )
    )
  ];
```
This looks absurdly complicated and it still doesn't work. The problem is that `coolercontrol.nix` needs at least the pkgs variable from it's own nixpkgs.
But still I get 
```
      error: attribute 'coolercontrol' missing

       at /nix/store/hrgdd7i21zacx7vjbqv40wknhhab7569-source/nixos/modules/programs/coolercontrol.nix:18:39:

           17|   config = lib.mkIf cfg.enable {
           18|     environment.systemPackages = with pkgs.coolercontrol; [
             |                                       ^
           19|       coolercontrol-gui

```

**Comment by OPNA2608** [comment]
> `flake.nix`

I don't use Flakes, so you may have to look up / ask in community channels about the specifics of how to do this in case it's different there, but perhaps you're lacking an overlay to add `coolercontrol` from `nixpkgs-coolercontrol`? With non-Flakes, this is prolly not perfect but evaluates & starts building everything:

```nix
let                                                                                                                      
  nixpkgs-coolercontrol-src = builtins.fetchTarball {                                                                    
    # https://github.com/codifryed/nixpkgs/tree/coolercontrol-0.17.0                                                     
    url = "https://github.com/codifryed/nixpkgs/archive/ed254575f639bc89d2b7856c603d46121f01bbc9.tar.gz";                
    sha256 = "09jkms9a4yikipmzpvmy3ybisw0chpf8zlp0ss9sch3262y72wkc";                                                     
  };                                                                                                                     
  nixpkgs-coolercontrol = import nixpkgs-coolercontrol-src { };                                                          
in {
  # ...
  imports = [
    # ...
    (nixpkgs-coolercontrol-src + "/nixos/modules/programs/coolercontrol.nix")
    # ...
  ];

  nixpkgs.overlays = [
    # Inherit `coolercontrol` attrset from coolercontrol PR
    (self: super: {
      inherit (nixpkgs-coolercontrol) coolercontrol;
    })
  ];

  programs.coolercontrol.enable = true;
```

**Comment by tbaumann** [comment]
Thanks, overlays was a concept I hadn't learned yet. But it should have made sense that this is what I should look into.

I will try that. Thanks

**Comment by tbaumann** [comment]
> ImportError: /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6: version `GLIBC_2.38' not found (required by /nix/store/zrq9b75k3sv0l45n66rz72c4h03i1dbw-glib-2.76.4/lib/libglib-2.0.so.0)
> ```

Nevermind that. I accidentally fucked around with LD_LIBRARY_PATH

**Comment by tbaumann** [comment]
This is potentially totally wrong and crazy. But this is how I got it to work

```nix
  imports = [
     (inputs.nixpkgs-coolercontrol + "/nixos/modules/programs/coolercontrol.nix")
  ];
  nixpkgs.overlays = [
    (self: super: {
      coolercontrol = lib.recurseIntoAttrs (super.callPackage ( inputs.nixpkgs-coolercontrol + /pkgs/applications/system/coolercontrol) { });
    })
  ];
  programs.coolercontrol.enable = true;
```

**Comment by magnetophon** [comment]
I was able to run the gui, but I couldn't control anything:
image

Maybe my HW is not supported?  I'm using a Framework laptop:
```
OS: NixOS 23.11pre549786.c757e9bd77b1 (Tapir) x86_64
Host: Framework FRANMACP04
Kernel: 6.5.11
Uptime: 2 days, 15 hours, 55 mins
Packages: 4207 (nix-system), 932 (nix-user), 35 (nix-default)
Shell: bash 5.2.15
Resolution: 2256x1504
DE: none+i3
WM: i3
Terminal: WezTerm
CPU: 12th Gen Intel i5-1240P (16) @ 1.700GHz
GPU: Intel Alder Lake-P GT2 [Iris Xe Graphics]
Memory: 19918MiB / 31816MiB
```

**Comment by tbaumann** [comment]
> I was able to run the gui, but I couldn't control anything: image
> 
> Maybe my HW is not supported? I'm using a Framework laptop:

I think Coolercontrol is mostly made to support devices controlled by liquidctl (Mostly NZXT RGB and fan controllers). Perhaps you have to enable hwmon support in the settings, I don't know what the default is.

**Comment by magnetophon** [comment]
Thanks for the quick reply.
```
coretemp-isa-0000
Adapter: ISA adapter
Package id 0:  +60.0°C  (high = +100.0°C, crit = +100.0°C)
Core 0:        +59.0°C  (high = +100.0°C, crit = +100.0°C)
Core 4:        +60.0°C  (high = +100.0°C, crit = +100.0°C)
Core 8:        +56.0°C  (high = +100.0°C, crit = +100.0°C)
Core 12:       +58.0°C  (high = +100.0°C, crit = +100.0°C)
Core 16:       +58.0°C  (high = +100.0°C, crit = +100.0°C)
Core 17:       +58.0°C  (high = +100.0°C, crit = +100.0°C)
Core 18:       +58.0°C  (high = +100.0°C, crit = +100.0°C)
Core 19:       +58.0°C  (high = +100.0°C, crit = +100.0°C)
Core 20:       +55.0°C  (high = +100.0°C, crit = +100.0°C)
Core 21:       +55.0°C  (high = +100.0°C, crit = +100.0°C)
Core 22:       +55.0°C  (high = +100.0°C, crit = +100.0°C)
Core 23:       +57.0°C  (high = +100.0°C, crit = +100.0°C)

BAT1-acpi-0
Adapter: ACPI interface
in0:          16.57 V
curr1:         5.00 mA

nvme-pci-0100
Adapter: PCI adapter
Composite:    +43.9°C  (low  =  -5.2°C, high = +89.8°C)
                       (crit = +93.8°C)

iwlwifi_1-virtual-0
Adapter: Virtual device
temp1:        +49.0°C

acpitz-acpi-0
Adapter: ACPI interface
temp1:        +62.8°C  (crit = +210.0°C)
temp2:        +55.8°C  (crit = +210.0°C)
temp3:        +58.8°C  (crit = +210.0°C)
temp4:        +54.8°C  (crit = +210.0°C)
temp5:        +34.8°C  (crit = +210.0°C)
```

BTW: Those last lines make me a bit nervous: ``crit = +210.0°C``.
Should I change anything?

**Comment by OPNA2608** [comment]
+1 to the above, I'm not sure if hwmon or liquidctl support your fans. From some light googling, you might need specialised software:

- https://community.frame.work/t/linux-fan-speed-controller-with-custom-speed-curve/17208
  - https://github.com/TamtamHero/fw-fanctrl
- https://www.howett.net/posts/2021-12-framework-ec
  - https://github.com/DHowett/framework-ec

**Comment by tbaumann** [comment]
> BTW: Those last lines make me a bit nervous: `crit = +210.0°C`. Should I change anything?

Many drivers don't come with sensible defaults. You could manually set those via https://linux.die.net/man/5/sensors.conf overrides. But it's not really worth it.

**Comment by magnetophon** [comment]
Thanks for the quick and informative replies!

**Review by magnetophon — APPROVED** [review]
Result of `nixpkgs-review pr 248972` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>coolercontrol.coolercontrol-gui</li>
    <li>coolercontrol.coolercontrol-gui.dist</li>
    <li>coolercontrol.coolercontrol-liqctld</li>
    <li>coolercontrol.coolercontrol-liqctld.dist</li>
    <li>coolercontrol.coolercontrold</li>
  </ul>
</details>

**Comment by OPNA2608** [comment]
[0.17.2 released](https://gitlab.com/coolercontrol/coolercontrol/-/tags/0.17.2), can we bump again?

**Comment by codifryed** [comment]
Hi @OPNA2608,  
Sure, I'll bump probably sometime this week. Additionally, coolercontrol may need a quick patch release in the interim.  

FYI, this release comes with an optional 'new UI preview' which needs an additional NPM build step before the coolercontrold cargo build. I probably won't include that in this nix build for simplicity right now, as I have quite of bit of development work to do for the project atm. 

Note that there are major build changes coming with 0.18.0 along with a new UI and additional features.

**Comment by codifryed** [comment]
Bumped and rebased as there have been a lot of changes over the last several months. 
@OPNA2608 can you double check that everything looks good from your side?

**Review by OPNA2608 — APPROVED** [review]
Sorry for the late reply, my desktop system with compatible hardware was down for some time.

image

LGTM.

**Comment by OPNA2608** [comment]
@ofborg build coolercontrol.coolercontrold coolercontrol.coolercontrold.passthru.tests coolercontrol.coolercontrol-liqctld coolercontrol.coolercontrol-liqctld.passthru.tests coolercontrol.coolercontrol-gui coolercontrol.coolercontrol-gui.passthru.tests

**Comment by codifryed** [comment]
Major update to CoolerControl 1.0.0 has hit with significant changes, including build ones. Will need some time to get the package working. In progress.

**Comment by codifryed** [comment]
I got it building again and have tested the daemons. 🎉 The UI I have difficulty testing though. (new dependencies and no more qt)

@OPNA2608 when you have a moment could you take a look?  
Also I'm not sure if I've done it the best way, but took some inspiration from other packages. Please also check if the UI is reachable on your machine from the daemon directly at http://localhost:11987/

**Comment by OPNA2608** [comment]
> http://localhost:11987/

Yep.
image

> desktop app

```
thread 'main' panicked at /build/cargo-vendor-dir/libappindicator-sys-0.7.3/src/lib.rs:41:5:
Failed to load ayatana-appindicator3 or appindicator3 dynamic library
libayatana-appindicator3.so.1: cannot open shared object file: No such file or directory
libappindicator3.so.1: cannot open shared object file: No such file or directory
libayatana-appindicator3.so: cannot open shared object file: No such file or directory
libappindicator3.so: cannot open shared object file: No such file or directory
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'll work on a fix in the packaging.

File: pkgs/applications/system/coolercontrol/coolercontrol-gui.nix
Comment (OPNA2608) [inline pkgs/applications/system/coolercontrol/coolercontrol-gui.nix:10]:
```suggestion
, libappindicator
, makeWrapper
```

File: pkgs/applications/system/coolercontrol/coolercontrol-gui.nix
Comment (OPNA2608) [inline pkgs/applications/system/coolercontrol/coolercontrol-gui.nix:48]:
```suggestion
    makeWrapper
    pkg-config
```

File: pkgs/applications/system/coolercontrol/coolercontrol-gui.nix
Comment (OPNA2608) [inline pkgs/applications/system/coolercontrol/coolercontrol-gui.nix:76]:
```suggestion

    wrapProgram $out/bin/coolercontrol \
      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ libappindicator ]}
  '';
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Suggestions should fix `libappindicator` finding at runtime.

Bildschirmfoto_2024-01-13_21-51-47


<details>
  <summary>Explanation, if desired</summary>

The library is in `buildInputs`, but that only matters for linking. `libappindicator` is not being linked, instead it gets looked up via `dlopen` at runtime.

The options for handling this are:

- Find the code that calls the equivalent of `dlopen("libname.so")` and `substituteInPlace` it to `dlopen("${lib.getLib libname}/lib/libname.so")`. But I have no clue how you'd patch that in vendored Rust dependencies
- Use `env.NIX_LDFLAGS = "-lname"` to force build-time linkage against the library, which will make `dlopen()` re-use that link if it has the same name. But this can be abit finnicky, and it's kind of an ugly solution IMO.
- Wrap the resulting binary to add the paths of required libraries to `LD_LIBRARY_PATH` at runtime before launching the real executable. That's what I've chosen here.

</details>

---

I'm not really a fan of the `ui-build` being copy-pasted in `coolercontrol-gui` and `coolercontrold`. I think it should be moved into a separate package like the other parts (smth like `coolercontrol-ui-data` maybe), and pulled in as an input into those parts that need it (`coolercontrol` at the top, then `coolercontrol.coolercontrol-ui-data` where needed).

Do you think you could do that, or any differing opinions on that?

(Thanks for all the effort you guys are pouring into this software btw, and your persistence with this PR :green_heart:)

**Comment by codifryed** [comment]
> Do you think you could do that, or any differing opinions on that?
> 
> (Thanks for all the effort you guys are pouring into this software btw, and your persistence with this PR 💚)

Thanks @OPNA2608 for your suggestions and working that out. 🙏🏼 

I saw several ways that other packages had dealt with libappindicator, and I didn't understand at first why a runtime dependency needed that level of build-customization. Your wrapper solution looks like the best by far from what I've seen though. 

I'll apply your suggested changes and also try to move the build-ui stuff out into its own thing. Oh, and a patch update is coming as well.

File: pkgs/applications/system/coolercontrol/coolercontrol-ui-data.nix
Comment (OPNA2608) [inline pkgs/applications/system/coolercontrol/coolercontrol-ui-data.nix:21]:
```suggestion

  meta = meta // {
    description = "${meta.description} (UI data)";
  };
}
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Small nit.

Do you want to keep bumping this until there's a period with less-frequent patch-level updates? FWIW I have merge permissions now, and this has been working fine for the last few months. Let me know when you think it's ready. :slightly_smiling_face:

**Comment by codifryed** [comment]
> Do you want to keep bumping this until there's a period with less-frequent patch-level updates? FWIW I have merge permissions now, and this has been working fine for the last few months. Let me know when you think it's ready. 🙂

What, you don't like regular updates? 😉 
Awesome news on the merge permissions. Yay! 🎉 

Explanation/Context:
One of the major challenges with software like this, as I'm sure you're probably aware, is the sheer number of hardware system variations that this software can support. One would think, that everything was all standardized and clear, but you'd be amazed at how much wiggle room there is from varying sources and sometimes each device. I think it's been functioning well for most people for quite some time, including myself. But there are some people & setups that haven't been so lucky. These quick patches right after the 1.0 release - which was a significant refactoring of both the front and backend - are to address some cases where it didn't work as it should with certain hardware setups. The majority of people I think haven't had many, if any, issues and there was quite extensive testing done before the release.
All of the actionable open issues have now been addressed, including those submitted and tested outside of GitLab, with the exception of your Desktop file issue which is planned for the next patch release. 😄  
If you'd prefer, we can wait for that one and then I'd say we're a go from my side.

**Comment by tbaumann** [comment]
Works for me, again.
I have been using the previous version for a long time. And since the wrapper fix the new one works really well too.

File: pkgs/applications/system/coolercontrol/coolercontrol-ui-data.nix
Comment (codifryed) [reply pkgs/applications/system/coolercontrol/coolercontrol-ui-data.nix:25]: applied.

**Comment by codifryed** [comment]
@OPNA2608  🙏🏼 
1.1.0 Update  
Release Notes: https://gitlab.com/coolercontrol/coolercontrol/-/releases/1.1.0  
We're OK-to-go from my side 👍🏼

**Comment by OPNA2608** [comment]
Been using the new version for a week, works fine (although the sudden password prompt & being denied access was very confusing until I read up on the changelog).

I think a final squash into just `coolercontrol.*: init at 1.1.0` (or individual, correctly-ordered `init`s for every derivation under `coolercontrol` if you want to be extra-proper) would be appropriate. I don't think all the intermediate fixups & bumps are of much use anymore once we merge, and some are likely covered by [this commit convention](https://github.com/NixOS/nixpkgs/blob/2565ab7b53ae6966bd08e68203a89645961849b3/CONTRIBUTING.md#commit-conventions):

> - If you have commits `pkg-name: oh, forgot to insert whitespace`: squash commits in this case. Use `git rebase -i`.

**Comment by codifryed** [comment]
> (although the sudden password prompt & being denied access was very confusing until I read up on the changelog)

That is not the intended behavior. It is supposed to auto-login for existing users that puts one into Admin mode right away. (Tested working ofc) There is a default password and changing it is optional. 

Does the Web UI also display this behavior for you, or just the Desktop App?

**Comment by OPNA2608** [comment]
> It is supposed to auto-login for existing users that puts one into Admin mode right away. (Tested working) There is a default password and changing it is optional.

On first reboot after update, it prompted me to authenticate with a password. Not knowing which password it was asking for, I tried 3 different ones. I assume that failed to authenticate, and the user menu in the top right said I was a guest afterwards.

I think I just relaunched the app, and that time it didn't prompt for a password and just logged me in as admin. Not knowing what suddenly made it not prompt for an unknown password, I quickly set a new one.

**Comment by codifryed** [comment]
Ok, thanks for the info @OPNA2608. 

I think the login-startup-timing for the desktop app then needs another tweak to avoid that. The desktop app in particular at startup is under a lot of load and can cause some noticeable delay until things have smoothed out. Understandably, it's unclear what to do in that situation. I'll add a bug report for others to know they can just click on 'login' afterwards if they experience this glitch, along with some more wiki info.

**Comment by codifryed** [comment]
@OPNA2608 

I tried to reproduce your issue on several systems, even under high load and couldn't get it to pop up. I have adjusted the login timeout so that the chance of that glitch happening again in the future should be near zero.

I squashed the commits and now there appears to be an upstream issue: The `liquidctl` build fails.
https://github.com/NixOS/nixpkgs/pull/288100

I'll keep an eye on that and push soon after that fix is through.

**Review by OPNA2608 — APPROVED** [review]
Result of `nixpkgs-review pr 248972` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>coolercontrol.coolercontrol-gui</li>
    <li>coolercontrol.coolercontrol-liqctld</li>
    <li>coolercontrol.coolercontrol-liqctld.dist</li>
    <li>coolercontrol.coolercontrol-ui-data</li>
    <li>coolercontrol.coolercontrold</li>
  </ul>
</details>

---

Sorry for the delay! Been using it on my system, works fine.

LGTM! Thanks for all the patience!

**Pull request merged**
